File: 2023/day05.cpp
// https://adventofcode.com/2023/day/5

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <array>
#include <print>
#include <ranges>

using range_t = aoc::closed_range<i64>;

struct single_mapping_t {
  range_t src;
  range_t dst;

  fn in_source_range(i64 value) const -> bool {
    return (value >= src.begin) && (value < src.end);
  }
};

struct Input {
  Vec<i64> seed_ints;
  Vec<Vec<single_mapping_t>> mappings;
};

auto parse(String const& filename) -> Input {
  Input result;
  std::ifstream file{filename};

  let line = aoc::read_line(file);
  result.seed_ints = aoc::split_to_vec<i64>(line.substr(sizeof("seeds:")), ' ');

  Vec<single_mapping_t> mapping;
  for (str line : aoc::views::read_lines(file)) {
    if (!aoc::is_number(line[0])) {
      stdr::sort(mapping, std::less<>{}, &single_mapping_t::src);
      result.mappings.push_back(std::move(mapping));
      mapping.clear();
      continue;
    }
    let[dest_start, source_start, range] =
        aoc::split<std::array<i64, 3>>(line, ' ');
    mapping.emplace_back(range_t{source_start, source_start + range},
                         range_t{dest_start, dest_start + range});
  }
  stdr::sort(mapping, std::less<>{}, &single_mapping_t::src);
  result.mappings.push_back(std::move(mapping));

  return result;
}

using seeds_t = Vec<range_t>;

fn map_single(range_t const& seed, single_mapping_t const& map) -> range_t {
  let offset = map.dst.begin - map.src.begin;
  return {seed.begin + offset, seed.end + offset};
}

// Diagram for the explanation of how an individual map applies to a seed range
// (x-y) seed range
// (a-b) map range
// Depending on the order, there are 4 possible scenarios
// Hyphen indicates copying the ranges as-is, plus means map is applied
// (x,a,b,y) - (x-a-1)(a+b)(b+1-y)
// (a,x,b,y) - (x+b)(b+1-y)
// (x,a,y,b) - (x-a-1)(a+y)
// (a,x,y,b) - (x+y)
fn apply_mapping(seeds_t current_seeds, Vec<single_mapping_t> const& mapping)
    -> seeds_t {
  // We want all the ranges to be sorted to simplify the algorithm
  AOC_ASSERT(stdr::is_sorted(mapping, std::less<>{}, &single_mapping_t::src),
             "mapping not presorted");

  seeds_t next_seeds;
  next_seeds.reserve(current_seeds.size());
  auto seed_it = std::begin(current_seeds);
  auto map_it = std::begin(mapping);

  auto find_relevant_map = [&]() {
    // Find first seed range where the map applies
    while (true) {
      if (map_it == std::end(mapping)) {
        // End of mapping, add all the remaining seeds as-is
        for (; seed_it != std::end(current_seeds); ++seed_it) {
          next_seeds.push_back(*seed_it);
        }
        break;
      }
      if (seed_it == std::end(current_seeds)) {
        // Out of seeds, ignore any remaining mapping
        break;
      }
      if (seed_it->overlaps_with(map_it->src)) {
        // Map applies to this seed range,
        // return to perform the rest of the algorithm
        break;
      }
      if (map_it->src < *seed_it) {
        // Map behind seed range
        ++map_it;
      } else {
        // Seed range behind map, pass seed range as-is
        next_seeds.push_back(*seed_it);
        ++seed_it;
      }
    }
  };

  auto split_seed_range = [&](const range_t& update_seed,
                              const range_t& new_seed) {
    *seed_it = update_seed;
    seed_it = current_seeds.insert(seed_it + 1, new_seed);
    ++map_it;
  };

  auto apply_map_to_seeds = [&]() {
    // Find last seed range where the map applies
    // For everything in between apply the map
    for (; seed_it != std::end(current_seeds); ++seed_it) {
      if (!seed_it->overlaps_with(map_it->src)) {
        // Map doesn't apply anymore, return to outer loop
        break;
      }
      // Note that end() points to 1 beyond the last element,
      // so we decrement by one
      let a = map_it->src.begin;
      let b = map_it->src.end - 1;
      let x = seed_it->begin;
      let y = seed_it->end - 1;
      if (seed_it->contains(map_it->src)) {
        // (x-a-1)(a+b)(b+1-y)
        if ((a - x) > 1) {
          // (x-a-1)
          next_seeds.emplace_back(x, a);
        }
        // (a+b)
        next_seeds.push_back(map_single(map_it->src, *map_it));
        if ((y - b) > 0) {
          // (b+1-y)
          split_seed_range({x, b + 2}, {b + 1, y + 1});
          // Important to break the loop after the split
          // because the map doesn't apply anymore
          break;
        }
      } else if (map_it->src.contains(*seed_it)) {
        // (a,x,y,b) - (x+y)
        next_seeds.push_back(map_single(*seed_it, *map_it));
      } else if ((x >= a) && (b <= y)) {
        // (a,x,b,y) - (x+b)(b+1-y)
        // (x+b)
        next_seeds.push_back(map_single({x, b + 1}, *map_it));
        // (b+1-y)
        split_seed_range({x, b + 2}, {b + 1, y + 1});
        // Important to break the loop after the split
        // because the map doesn't apply anymore
        break;
      } else {
        // (x,a,y,b) - (x-a-1)(a+y)
        // (x-a-1)
        next_seeds.emplace_back(x, a);
        // (a+y)
        next_seeds.push_back(map_single({a, y + 1}, *map_it));
      }
    }
  };

  while (true) {
    find_relevant_map();
    if ((map_it == std::end(mapping)) || (seed_it == std::end(current_seeds))) {
      break;
    }
    apply_map_to_seeds();
  }

  stdr::sort(next_seeds, std::less<>{});
  return next_seeds;
}

template <bool full_ranges>
fn solve_case(Input const& input) -> i64 {
  seeds_t current_seeds;
  for (usize i = 0; i + 1 < input.seed_ints.size(); i += 2) {
    let v0 = input.seed_ints[i];
    let v1 = input.seed_ints[i + 1];
    if constexpr (!full_ranges) {
      current_seeds.emplace_back(v0, v0 + 1);
      current_seeds.emplace_back(v1, v1 + 1);
    } else {
      current_seeds.emplace_back(v0, v0 + v1);
    }
  }

  for (let& mapping : input.mappings) {
    current_seeds = apply_mapping(current_seeds, mapping);
  }

  return current_seeds[0].begin;
}

int main() {
  std::println("Part 1");
  let example = parse("day05.example");
  AOC_EXPECT_RESULT(35, (solve_case<false>(example)));
  let input = parse("day05.input");
  AOC_EXPECT_RESULT(486613012, (solve_case<false>(input)));

  std::println("Part 2");
  AOC_EXPECT_RESULT(46, (solve_case<true>(example)));
  AOC_EXPECT_RESULT(56931769, (solve_case<true>(input)));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2023/src/day05.rs
type Segment = aoc::closed_range::ClosedRange<i64>;

#[derive(Clone, Copy)]
struct SingleMapping {
    src: Segment,
    dst: Segment,
}

struct Input {
    seed_ints: Vec<i64>,
    mappings: Vec<Vec<SingleMapping>>,
}

fn parse(filename: &str) -> Input {
    let lines = aoc::file::read_lines(filename);
    let mut iter = lines.iter();

    let first_line = iter.next().unwrap();
    let (_, seeds_str) = first_line.split_once(':').unwrap();
    let seed_ints: Vec<i64> = seeds_str
        .split_whitespace()
        .map(|n| n.parse().unwrap())
        .collect();

    let mut mappings: Vec<Vec<SingleMapping>> = Vec::new();
    let mut mapping: Vec<SingleMapping> = Vec::new();

    for line in iter {
        if !line.starts_with(|c: char| c.is_ascii_digit()) {
            mapping.sort_unstable_by_key(|m| m.src);
            mappings.push(std::mem::take(&mut mapping));
            continue;
        }
        let mut parts = line.split_whitespace();
        let dest_start: i64 = parts.next().unwrap().parse().unwrap();
        let source_start: i64 = parts.next().unwrap().parse().unwrap();
        let range: i64 = parts.next().unwrap().parse().unwrap();
        mapping.push(SingleMapping {
            src: Segment::new(source_start, source_start + range),
            dst: Segment::new(dest_start, dest_start + range),
        });
    }
    mapping.sort_unstable_by_key(|m| m.src);
    mappings.push(mapping);

    Input {
        seed_ints,
        mappings,
    }
}

fn map_single(seed: Segment, map: SingleMapping) -> Segment {
    let offset = map.dst.begin - map.src.begin;
    Segment::new(seed.begin + offset, seed.end + offset)
}

// Diagram for the explanation of how an individual map applies to a seed range
// (x-y) seed range
// (a-b) map range
// Depending on the order, there are 4 possible scenarios
// Hyphen indicates copying the ranges as-is, plus means map is applied
// (x,a,b,y) - (x-a-1)(a+b)(b+1-y)
// (a,x,b,y) - (x+b)(b+1-y)
// (x,a,y,b) - (x-a-1)(a+y)
// (a,x,y,b) - (x+y)
fn apply_mapping(mut current_seeds: Vec<Segment>, mapping: &[SingleMapping]) -> Vec<Segment> {
    // We want all the ranges to be sorted to simplify the algorithm
    debug_assert!(
        mapping.windows(2).all(|w| w[0].src <= w[1].src),
        "mapping not presorted"
    );

    let mut next_seeds: Vec<Segment> = Vec::with_capacity(current_seeds.len());
    let mut seed_idx = 0usize;
    let mut map_idx = 0usize;

    loop {
        // find_relevant_map: find first seed range where the map applies
        loop {
            if map_idx >= mapping.len() {
                // End of mapping, add all the remaining seeds as-is
                next_seeds.extend_from_slice(&current_seeds[seed_idx..]);
                seed_idx = current_seeds.len();
                break;
            }
            if seed_idx >= current_seeds.len() {
                // Out of seeds, ignore any remaining mapping
                break;
            }
            if current_seeds[seed_idx].overlaps_with(mapping[map_idx].src) {
                // Map applies to this seed range,
                // return to perform the rest of the algorithm
                break;
            }
            if mapping[map_idx].src < current_seeds[seed_idx] {
                // Map behind seed range
                map_idx += 1;
            } else {
                // Seed range behind map, pass seed range as-is
                next_seeds.push(current_seeds[seed_idx]);
                seed_idx += 1;
            }
        }

        if map_idx >= mapping.len() || seed_idx >= current_seeds.len() {
            break;
        }

        // apply_map_to_seeds: find last seed range where the map applies,
        // for everything in between apply the map
        'apply: loop {
            if seed_idx >= current_seeds.len() {
                break;
            }
            if !current_seeds[seed_idx].overlaps_with(mapping[map_idx].src) {
                // Map doesn't apply anymore, return to outer loop
                break;
            }
            // Note that end() points to 1 beyond the last element,
            // so we decrement by one
            let a = mapping[map_idx].src.begin;
            let b = mapping[map_idx].src.end - 1;
            let x = current_seeds[seed_idx].begin;
            let y = current_seeds[seed_idx].end - 1;

            let split = if current_seeds[seed_idx].contains(mapping[map_idx].src) {
                // (x-a-1)(a+b)(b+1-y)
                if (a - x) > 1 {
                    // (x-a-1)
                    next_seeds.push(Segment::new(x, a));
                }
                // (a+b)
                next_seeds.push(map_single(mapping[map_idx].src, mapping[map_idx]));
                if (y - b) > 0 {
                    // (b+1-y)
                    current_seeds[seed_idx] = Segment::new(x, b + 2);
                    current_seeds.insert(seed_idx + 1, Segment::new(b + 1, y + 1));
                    seed_idx += 1;
                    map_idx += 1;
                    true
                } else {
                    false
                }
            } else if mapping[map_idx].src.contains(current_seeds[seed_idx]) {
                // (a,x,y,b) - (x+y)
                next_seeds.push(map_single(current_seeds[seed_idx], mapping[map_idx]));
                false
            } else if (x >= a) && (b <= y) {
                // (a,x,b,y) - (x+b)(b+1-y)
                // (x+b)
                next_seeds.push(map_single(Segment::new(x, b + 1), mapping[map_idx]));
                // (b+1-y)
                current_seeds[seed_idx] = Segment::new(x, b + 2);
                current_seeds.insert(seed_idx + 1, Segment::new(b + 1, y + 1));
                seed_idx += 1;
                map_idx += 1;
                true
            } else {
                // (x,a,y,b) - (x-a-1)(a+y)
                // (x-a-1)
                next_seeds.push(Segment::new(x, a));
                // (a+y)
                next_seeds.push(map_single(Segment::new(a, y + 1), mapping[map_idx]));
                false
            };

            if split {
                // Important to break the loop after the split
                // because the map doesn't apply anymore
                break 'apply;
            }
            seed_idx += 1;
        }
    }

    next_seeds.sort_unstable();
    next_seeds
}

fn solve_case1(input: &Input) -> i64 {
    let mut current_seeds: Vec<Segment> = input
        .seed_ints
        .chunks(2)
        .flat_map(|c| [Segment::new(c[0], c[0] + 1), Segment::new(c[1], c[1] + 1)])
        .collect();
    for mapping in &input.mappings {
        current_seeds = apply_mapping(current_seeds, mapping);
    }
    current_seeds[0].begin
}

fn solve_case2(input: &Input) -> i64 {
    let mut current_seeds: Vec<Segment> = input
        .seed_ints
        .chunks(2)
        .map(|c| Segment::new(c[0], c[0] + c[1]))
        .collect();
    for mapping in &input.mappings {
        current_seeds = apply_mapping(current_seeds, mapping);
    }
    current_seeds[0].begin
}

fn main() {
    println!("Part 1");
    let example = parse("day05.example");
    aoc::expect_result!(35, solve_case1(&example));
    let input = parse("day05.input");
    aoc::expect_result!(486613012, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(46, solve_case2(&example));
    aoc::expect_result!(56931769, solve_case2(&input));
}
